src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java

```diff
@@ -56,4 +101,92 @@ public static String getTopicCmekKey(String topicId) throws IOException {
     return getTopic(topicId).getKmsKeyName();
   }
 
+
+  public static void publishWithErrorHandlerExample(String projectId, String topicId, List<String> dataMessages)
+    throws IOException, InterruptedException {
+    TopicName topicName = TopicName.of(projectId, topicId);
+    Publisher publisher = null;
+    try {
+      publisher = Publisher.newBuilder(topicName).build();
+      for (final String message : dataMessages) {
+        ByteString data = ByteString.copyFromUtf8(message);
+        PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(data).build();
+        ApiFuture<String> future = publisher.publish(pubsubMessage);
+
+        /**
+         * Adding an asynchronous callback to handle success / failure.
+         */
+        ApiFutures.addCallback(
+          future,
+          new ApiFutureCallback<String>() {
+
+            @Override
+            public void onFailure(Throwable throwable) {
+              if (throwable instanceof ApiException) {
+                ApiException apiException = ((ApiException) throwable);
+                /**
+                 * details on the API exception.
+                 */
+                logger.info(String.valueOf(apiException.getStatusCode().getCode()));
+                logger.info(String.valueOf(apiException.isRetryable()));
+              }
+              logger.info("Error publishing message : " + message);
+            }
+
+            @Override
+            public void onSuccess(String messageId) {
+              /**
+               * Once published, returns server-assigned message ids (unique within the topic).
+               */
+              logger.info("Published message ID: " + messageId);
+            }
+          },
+          MoreExecutors.directExecutor());
+      }
+    } finally {
+      if (publisher != null) {
+        /**
+         * When finished with the publisher, shutdown to free up resources.
+         */
+        publisher.shutdown();
+        publisher.awaitTermination(1, TimeUnit.MINUTES);
+      }
+    }
+  }
+
+  public static void subscribeAsyncExample(String projectId, String subscriptionId) {
+    ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(PluginPropertyUtils.pluginProp
+      (ConstantsUtil.PROJECT_ID), subscriptionId);
```

Comment:
use projectId from the function parameter

Reply:
changed